wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,683 @@
+-- not for generating truly random Mnemonic words.
+--
+-- See 'Crypto.Random.Entropy (getEntropy)'
+instance (KnownNat n, SupportedWords n) => Arbitrary (Entropy n) where
+    arbitrary =
+        let
+            size    = entropyLength (Proxy :: Proxy (Entropy n))
+            entropy = mkEntropy . B8.pack <$> vectorOf (size `quot` 8) arbitrary
+        in
+            either (error . ("Invalid Arbitrary Entropy: " <>)) identity <$> entropy
+
+
+-- Same remark from 'Arbitrary Entropy' applies here.
+instance (KnownNat n, SupportedWords n) => Arbitrary (Mnemonic n) where
+    arbitrary =
+        entropyToMnemonic <$> arbitrary
+
+
+-- FIXME: Suggestion, we could -- when certain flags are turned on -- display
+-- a fingerprint of the Mnemonic, like a PKBDF2 over n iterations. This could be
+-- useful for debug to know whether two users are using the same mnemonic words
+-- and relatively benign EVEN THOUGH, it will permit to somewhat tight requests
+-- to a specific identity (since mnemonic words are 'unique', they are supposed
+-- to uniquely identify users, hence the privacy issue). For debbugging only and
+-- with the user consent, that's something we could do.
+instance Buildable (Mnemonic n) where
+    build _ =
+        "<mnemonic>"
+
+instance Buildable (SecureLog (Mnemonic n)) where
+    build _ =
+        "<mnemonic>"
+
+-- | To use everytime we need to show an example of a Mnemonic. This particular
+-- mnemonic is rejected to prevent users from using it on a real wallet.
+instance Default (Mnemonic 12) where
+    def = Mnemonic
+        { getEntropy = Entropy "\211\177\US\"\245\163\233\152\169]\149\DC3\239)\234\172"
+        , getWords =
+            [ (1693,"squirrel")
+            , (1095,"material")
+            , (1605,"silly")
+            , (1882,"twice")
+            , (500,"direct")
+            , (1634,"slush")
+            , (1323,"pistol")
+            , (1429,"razor")
+            , (159,"become")
+            , (970,"junk")
+            , (981,"kingdom")
+            , (710,"flee")
+            ]
+        }
+
+instance (SupportedWords n, KnownNat n) => FromJSON (Mnemonic n) where
+    parseJSON =
+        parseJSON >=> (eitherToParser . mkMnemonic)
+      where
+        eitherToParser :: Either Text (Mnemonic n) -> Parser (Mnemonic n)
+        eitherToParser =
+            either (fail . show) pure
+
+instance (SupportedWords n, KnownNat n) => ToJSON (Mnemonic n) where
+    toJSON =
+        toJSON . map snd . getWords
+
+instance (SupportedWords n, KnownNat n) => ToSchema (Mnemonic n) where
+    declareNamedSchema _ = do
+        let n = natVal (Proxy :: Proxy n)
+        NamedSchema _ schema <- declareNamedSchema (Proxy @[Text])
+        return $ NamedSchema (Just "Mnemonic") schema
+            & minItems ?~ n
+            & maxItems ?~ n
+
+
+-- | Smart-constructor for the Entropy
+mkEntropy
+    :: forall n. (SupportedWords n, KnownNat n)
+    => ByteString
+    -> Either Text (Entropy n)
```

Comment:
Instead of `Text`, can we make a datatype that represents the failure? Seems like:

```haskell
data EntropyError = EntropyError
  { requiredBytes :: Int
  , providedBytes :: Int
  }
```

oculd cover it

Reply:
Good idea :)